### PR TITLE
Bug 1392855 cliqz funnelcake experiment

### PR DESCRIPTION
**Pull request by alexgibson**
## Description
- Adds 4 funnelcake experiments for bundling the Cliqz extension in Firefox.
- Experiment is targeted for the `de` locale and for geolocation in Germany.
- ~Final requirements are still being ironed out, so this is **do not merge** for now.~

## Issue / Bugzilla link
https://bugzilla.mozilla.org/show_bug.cgi?id=1392855

## Testing
Demo `/new/`:

- https://bedrock-demo-agibson.us-west.moz.works/de/firefox/new/?f=119
- https://bedrock-demo-agibson.us-west.moz.works/de/firefox/new/?f=120
- https://bedrock-demo-agibson.us-west.moz.works/de/firefox/new/?f=121 (cliqz privacy policy)
- https://bedrock-demo-agibson.us-west.moz.works/de/firefox/new/?f=122 (cliqz privacy policy)

Demo '/firstrun/':

- https://bedrock-demo-agibson.us-west.moz.works/de/firefox/56.0/firstrun/?f=119 (control)
- https://bedrock-demo-agibson.us-west.moz.works/de/firefox/56.0/firstrun/?f=120 (firefox template)
- https://bedrock-demo-agibson.us-west.moz.works/de/firefox/56.0/firstrun/?f=121 (cliqz template)
- https://bedrock-demo-agibson.us-west.moz.works/de/firefox/56.0/firstrun/?f=122 (cliqz template)

### CODE REVIEW

File: jenkins/branches/demo/agibson.env [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#funnelcakes
```
Comment (alexgibson) [inline jenkins/branches/demo/agibson.env]: Reminder for me to remove this

File: CLIQZ.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+## Audience Targeting
```
Comment (alexgibson) [inline CLIQZ.md]: Reminder for me to remove this file before merging.

File: media/css/firefox/firstrun/firstrun-cliqz-funnelcake.less [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+#content {
+    padding: 0 100px;
+    text-align: center;
+    color: #fff;
+    position: relative;
+    z-index: 5;
+
+    h1 {
+        font-size: 42px;
```
Comment (schalkneethling) [inline media/css/firefox/firstrun/firstrun-cliqz-funnelcake.less]: Should this be using the `.font-size` mixin?

File: media/css/firefox/firstrun/firstrun-cliqz-funnelcake.less [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        float: left;
+        margin: 0;
+        padding: 0;
+        width: 50%;
+    }
+
+    h2 {
+        margin-bottom: 10px;
+        padding-top: 10px;
+        font-size: 20px;
```
Comment (schalkneethling) [inline media/css/firefox/firstrun/firstrun-cliqz-funnelcake.less]: Should this be using the `.font-size` mixin?

File: media/css/firefox/firstrun/firstrun-cliqz-funnelcake.less [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    }
+
+    p {
+        margin-top: 0;
+        padding: 0 48px;
+        font-size: 18px;
```
Comment (schalkneethling) [inline media/css/firefox/firstrun/firstrun-cliqz-funnelcake.less]: Should this be using the `.font-size` mixin?

File: media/js/base/mozilla-traffic-cop-funnelcake-geo.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        xhr.open('GET', '/country-code.json');
+        // must come after open call above for IE 10 & 11
+        xhr.timeout = 2000;
+        xhr.send();
+    };
+
+    // starts the traffic cop experiment
+    TCFCGeoExp.runExperiment = function(config) {
+        var rawls = new Mozilla.TrafficCop(config);
+
```
Comment (schalkneethling) [inline media/js/base/mozilla-traffic-cop-funnelcake-geo.js]: Nit: Is this empty line intentional?

File: media/js/firefox/new/experiment-cliqz-funnelcakes.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+(function(Mozilla) {
+    'use strict';
+
+    // initialize experiment - performs env check and does geo-lookup
+    Mozilla.TCFCGeoExp.init({
+        countryCode: 'us',
```
Comment (schalkneethling) [inline media/js/firefox/new/experiment-cliqz-funnelcakes.js]: Should this be 'de'?

File: tests/unit/spec/base/mozilla-traffic-cop-funnelcake-geo.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+describe('mozilla-traffic-cop-funnelcake-geo.js', function() {
+    'use strict';
+
+    describe('checkInCohort', function() {
+        it('should return true if user has the provided cookie', function() {
+            spyOn(Mozilla.Cookies, 'hasItem').and.returnValue(true);
+            expect(Mozilla.TCFCGeoExp.checkInCohort('foo')).toBeTruthy();
+        });
+
+        it('should return false if user does not has the provided cookie', function() {
```
Comment (schalkneethling) [inline tests/unit/spec/base/mozilla-traffic-cop-funnelcake-geo.js]: Nit: should return false if user does not `have` the provided cookie

File: tests/unit/spec/base/mozilla-traffic-cop-funnelcake-geo.js [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        });
+    });
+
+    describe('checkGeoNonmatch', function() {
+        it('should return false if user has the provided cookie', function() {
+            spyOn(Mozilla.Cookies, 'hasItem').and.returnValue(true);
+            expect(Mozilla.TCFCGeoExp.checkGeoNonmatch('foo')).toBeFalsy();
+        });
+
+        it('should return true if user does not has the provided cookie', function() {
```
Comment (schalkneethling) [inline tests/unit/spec/base/mozilla-traffic-cop-funnelcake-geo.js]: Nit: should return true if user does not `have` the provided cookie

File: media/js/firefox/new/experiment-cliqz-funnelcakes.js (hunk 7 above)
Comment (alexgibson) [reply media/js/firefox/new/experiment-cliqz-funnelcakes.js]: Yeah, this has just been left in for testing purposes

**Comment by alexgibson** [comment]
@jpetto - ok, I think the final bits of this are finally in place. When you're back, it would be good to get a second review on what's here just in case I've missed anything.

File: media/js/firefox/new/experiment-cliqz-funnelcakes.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    // initialize experiment - performs env check and does geo-lookup
+    Mozilla.TCFCGeoExp.init({
+        countryCode: 'de',
+        experimentConfig: {
+            id: 'experiment_cliqz_funnelcake',
+            variations: {
+                'f=119': 25,
+                'f=120': 25,
+                'f=121': 25,
+                'f=122': 25
```
Comment (jpetto) [inline media/js/firefox/new/experiment-cliqz-funnelcakes.js]: Just verifying these percentages are set (didn't see anything specific in the bug), or are known to need updating prior to merge.

File: media/js/base/mozilla-traffic-cop-funnelcake-geo.js [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    };
+
+    // ensures current visitor did not previously fail the geo-lookup
+    TCFCGeoExp.checkGeoNonmatch = function(geoNonmatchCookieName) {
+        // check if user already failed geolookup
+        return !Mozilla.Cookies.hasItem(geoNonmatchCookieName);
+    };
+
+    // checks to see if visitor was previously entered into a variation
+    TCFCGeoExp.checkInCohort = function(hasCookies, experimentId) {
```
[the remark this answers is not in the document]
Comment (alexgibson) [reply media/js/base/mozilla-traffic-cop-funnelcake-geo.js]: Nice spot 👍

File: media/js/firefox/new/experiment-cliqz-funnelcakes.js (hunk 10 above)
Comment (alexgibson) [reply media/js/firefox/new/experiment-cliqz-funnelcakes.js]: Yes these are final (confirmed in the solidarity that is Slack, alas).

**Comment by alexgibson** [comment]
Thanks @jpetto - updated and rebased. The latest email thread says they want to launch this experiment on Monday, so if all looks good I think we should probably merge & push this today, so it can sit until ready. We could enable things on stage on Friday for testing?

**Comment by DanielRuf** [comment]
Not a great idea imho.

**Pull request merged**